### ISSUE TITLE
Suggestion: Use an MSYS2 package

### ISSUE THREAD

**Issue by DavidEGrayson** [ARMmbed/yotta_windows_installer#13]
This isn't really an issue, just a suggestion: you could use [MSYS2](https://msys2.github.io/) packages to distribute yotta and its dependencies instead of this NSIS installer.  MSYS2 is a really nice distribution and package manager for open source software in Windows, and it already has cmake, make, ninja, and Python 2 packages.  There is an arm-none-eabi-gcc package, but there might be some problems with it because it is not being distributed at http://repo.msys2.org/mingw/x86_64/ currently.

The benefits would be:

* With MSYS2, Windows becomes less of a special case among operating systems because MSYS2 comes with a bash shell and has packages for all the usual Unix utilities that people expect (like grep).
* Users can pick and choose which packages to install; if they don't need ARM gcc because they are doing AVR stuff, maybe they wouldn't install the ARM gcc package.
* MSYS2 packages feel more open source than most open source software because each one comes with a publicly-available PKGBUILD script that is easy to find and it has the exact commands needed to build the software.  Here is the script for ninja: https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-ninja/PKGBUILD
* Users can conveniently receive updates to the software by running a few shell commands (`update-core` and `pacman -Su`).

Every official native Windows MSYS2 package has build scripts which are found here: https://github.com/Alexpux/MINGW-packages .  The easiest option is to make build scripts for your software and submit them as pull requests to that repository.  Alternatively, you could host the binary packages on your own server, and users of MSYS2 could manually install a package from your server that adds your server to the list of servers that pacman gets packages from.  Then they would be able to install your software with something like `pacman -Sy && pacman -S mingw-w64-x86_64-yotta`.  To make it easier for beginners that don't already have MSYS2, you could provide your own MSYS2 installer where all the needed packages are already installed and pacman is already configured.

I mentioned this in a [YouTube comment](https://www.youtube.com/watch?v=4fqiHUyyXvU&lc=z124wrejakrxfxjth23rtjrxzqrudhqhy04) and I was encouraged to open an issue here instead.

**Comment by BlackstoneEngineering** [comment]
@DavidEGrayson That sounds AWESOME!! Is MYSYS2 using the pacman package manager same as ARCH does? Also, would you be interested in spear-heading this project and getting it going? I must admit I am out of my depth and have never worked with the MYSYS2 platform before.

**Comment by DavidEGrayson** [comment]
I am glad you like the suggestion!  Yes, it uses the same package manager as Arch Linux, with some patches to make it work on Windows.

Information about how to build MSYS2 packages is here on their wiki:
https://sourceforge.net/p/msys2/wiki/Contributing%20to%20MSYS2/

I enjoy working on MSYS2 packages, so yes, I suppose I can spearhead this one if you are enthusiastic about it.  I will only be working on it in my free time one or two times per week, though, so don't expect me to meet any particular deadline.  The biggest problem right now is that MSYS2 does not have compiled packages for arm-none-eabi-gcc, or the matching binutils and C library.  I'll try to figure out why that is and what I can do about it.

**Comment by BlackstoneEngineering** [comment]
Awesome-sauce! Please take the lead, and I totally understand about the time commitment, anything you can contribute would be super-badass-amazing. 

I'm going to do a deep dive on this as soon as time permits.

Looking forward to building awesome with you mate!

**Comment by BlackstoneEngineering** [comment]
@DavidEGrayson So i took a look at msys2 and i'm having a problem with the cryptography package. I cant seem to get it to work, specifically the `pip install -U cryptography` option. Have you had any luck? 

I managed to get yotta's dependencies installed, 
* gcc
* cmake
* ninja
* python

but I cant get yotta to install due to a missing ffi.h file from what I suspect is the libssl-dev package that it requires.

**Comment by DavidEGrayson** [comment]
I'll see if I can get the cryptography pip to install on my machine.  To get ffi.h, you should install `mingw-w64-x86_64-libffi`.  More generally, you can run `pkgfile --update` and `pkgfile ffi.h` to find all packages that provide a file by that name.

I haven't done anything related to the toolchain yet but I'm hoping to look into it soon.

**Comment by DavidEGrayson** [comment]
When I do `pip install -U cryptography` in MSYS2, I get [this output](https://gist.github.com/DavidEGrayson/b333c598fedc8db8397b), with error message "VC 6.0 is not supported by this module".  The problem is that it's trying to MSVC (Microsoft's C compiler).  Really it should just use the gcc.exe that is on the path.

**Comment by DavidEGrayson** [comment]
P.S.  I am glad that you are looking in to MSYS2!  :+1:

**Comment by BlackstoneEngineering** [comment]
[Heres what I got](https://gist.github.com/BlackstoneEngineering/9e1f4613c2401ad1a562) as output from my screen. I've got the nagging feeling i've been here before and solved this on my cygwin install on a different machine, cant remember for the life of me how i resolved it though...

**Comment by DavidEGrayson** [comment]
Okay, I can see from your shell session that MSYSTEM is "MSYS", which means you are compiling software that depends on msys-2.0.dll, which provides POSIX emulation.  I would encourage you to compiled native Windows software (mingw-w64) instead unless there is some component that really can't be ported over to MinGW.  This will probably make the software faster, and if someone ever wants a standalone installer without MSYS2 it should be easier to do that.  So you would run the MSYS2 Win64 shell from the Start Menu and install `mingw-w64-x86_64-gcc`, `mingw-w64-x86_64-ninja`, and packages like that.  Once you install `mingw-w64-x86_64-libffi`, that should resolve the error message about ffi.h not being found.

**Comment by DavidEGrayson** [comment]
I was able to solve my "VC 6.0" my commenting out the two lines related to msvc9_support in /mingw64/lib/python2.7/site-packages/setuptools/extension.py.  I should make a pull request to https://github.com/Alexpux/MINGW-packages to fix that problem.  My current problem is that compiling the pyOCD pip takes many minutes, hogs the CPU, and eventually causes my system to run out of memory.  That might just be my personal problem though.  :imp:

**Comment by DavidEGrayson** [comment]
I didn't realize that yotta depends on so many pips.  How do those pips get compiled with your NSIS installer?  Does the user need to have Visual Studio installed, or do those pips come with precompiled binaries or something?

**Comment by BlackstoneEngineering** [comment]
its the darndest thing, even though i have mingw-64 python2 installed I cant hit it from the command line, its like the path to it isnt set when its installed. I'm going to try this on a sandboxed machine tomorrow, its looking like my system path isnt getting modified by msys2 so its not finding the c:/msys64/mingw64/bin folder.

To answer your question about the pips, i have no clue, i just know if i install the pre-reqs everything else takes care of itself, yay pip.

@autopulated : care to weigh in here?

**Comment by DavidEGrayson** [comment]
Well, /mingw64/bin should get added to your path by mingw64_shell.bat.  You can type `echo $PATH` in bash to see if it's really on your path.  Running `hash -r` might also help.

**Comment by BlackstoneEngineering** [comment]
Thats just it, it doesn't show up in msys2

```shell
ImaUsr@TheMachine MSYS ~
$ echo $PATH
/usr/local/bin:/usr/bin:/bin:/opt/bin:/c/PROGRAMDATA/ORACLE/JAVA/JAVAPATH:/c/PROGRAM FILES (X86)/INTEL/ICLS CLIENT:/c/PROGRAM FILES/INTEL/ICLS CLIENT:/c/PROGRAM FILES (X86)/NVIDIA CORPORATION/PHYSX/COMMON:/c/Windows/SYSTEM32:/c/Windows:/c/Windows/SYSTEM32/WBEM:/c/Windows/SYSTEM32/WINDOWSPOWERSHELL/V1.0:/c/PROGRAM FILES/INTEL/INTEL(R) MANAGEMENT ENGINE COMPONENTS/DAL:/c/PROGRAM FILES/INTEL/INTEL(R) MANAGEMENT ENGINE COMPONENTS/IPT:/c/PROGRAM FILES (X86)/INTEL/INTEL(R) MANAGEMENT ENGINE COMPONENTS/DAL:/c/PROGRAM FILES (X86)/INTEL/INTEL(R) MANAGEMENT ENGINE COMPONENTS/IPT:/c/Program Files/WIDCOMM/Bluetooth Software:/c/Program Files/WIDCOMM/Bluetooth Software/syswow64

```

**Comment by DavidEGrayson** [comment]
Did you start the shell mingw64_shell.bat or with the corresponding entry in the Start Menu that says something like "Win64 shell"?

**Comment by BlackstoneEngineering** [comment]
I ran the shortcut from the start menu that runs this command 
```C:\Windows\System32\cmd.exe /A /Q /K C:\msys64\msys2_shell.bat```

**Comment by DavidEGrayson** [comment]
OK, you should try to use `mingw64_shell.bat` or the corresponding Start Menu entry instead.

**Comment by BlackstoneEngineering** [comment]
OoOoOoOo gotcha, different shells are provided, kk, tried that one, still failing yotta install. going to bed now, will take another whack at this tomorrow with my VM at work. Thank you for all you

**Comment by BlackstoneEngineering** [comment]
Whoops, my mouse slipped, wasnt supposed to close that, srry, my bad

**Comment by DavidEGrayson** [comment]
Yeah, the basic idea is that the different shells have a different gcc.exe and a different set of programs on their path, which helps you make software for different Windows runtime environments (msys-2.0.dll, win32, or win64).

**Comment by BlackstoneEngineering** [comment]
ah, that probably explains my issues then, will re-try with the proper shell tomorrow

**Comment by DavidEGrayson** [comment]
Fixing the "VC 6.0" is harder than I was expecting.  If I reinstall setuptools by running `pacman -S mingw-w64-x86_64-python2-setuptools` then I can see that /mingw64/lib/python2.7/site-packages/setuptools/extension.py doesn't have the offending line in it.  Then when I run `pip -U cryptography` it installs setuptools itself, which brings in a new version of setuptools that has those offending lines in it.

The general issue here is that pacman is a package manager and its packages are maintained by a certain set of people that know how MSYS2 works, and pip is also a package manager and it is perhaps maintained by people that don't care about MSYS2 or test with it regularly.

One solution might be to update the mingw-w64-x86_64-python2-setuptools package so that pip views it as up-to-date and doesn't try to update it from the internet.  Of course, this will stop working as soon as a new version of setuptools comes out, but that might be OK.

**Comment by DavidEGrayson** [comment]
You might want to try running `update-core` followed by `pacman -Su` tomorrow to get the latest version of setuptools (18.4).  This is assuming that @AlexPux rebuilds the package tonight.  Also, you might try installing the cryptography pip without using the `-U` option, because it is bad for pip to update the python setuptools package in MSYS2.

**Comment by autopulated** [comment]
all of yotta's dependencies have "wheel" pre-compiled packages for windows –  though these will only work with standard python.

I think that https://github.com/project-generator/project_generator has just released a new version (0.6) which drops the dependency on cython, which is one of the painful ones, so maybe that will help.

**Comment by DavidEGrayson** [comment]
@BlackstoneEngineering I asked about the status of arm-none-eabi-gcc in MSYS2 here: https://github.com/Alexpux/MINGW-packages/issues/842  Since MSYS2 has upgraded to GCC 5, they aren't really interested in making a working ARM toolchian until it too can be upgraded to GCC 5.  ARM's GCC 5.0 toolchain will probably be available by the end of the year.

**Comment by BlackstoneEngineering** [comment]
KK, for the moment I just went ahead and installed arm-none-eabi-gcc with the standalone windows installer. I have tried a couple different ways of instalilng dependencies, tried using just msys packages, tried just using mingw64 packages, tried a mixture, and i have yet to be able to get a working yotta install downloaded. Most of the issues i'm seeing seem to be around the crypto library and the openssl library. @DavidEGrayson is there any chance I could ask you to give installing yotta s hot with MSYS2 to see if you can get it to work?

**Comment by DavidEGrayson** [comment]
I tried to get the python [cryptography](https://pypi.python.org/pypi/cryptography/0.2.1) package to install.  Most of [cryptography's dependencies](https://gist.github.com/DavidEGrayson/007688ce94220048b07c) can be installed fine with either pacman or pip, but the cffi dependency had an error compiling.  So I tried to make a cffi package, but the tests are failing with this output when I run them:

~~~~
$ py.test c/ testing/
============================= test session starts =============================
platform win32 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: D:/builds/mingw-packages/mingw-w64-python-cffi/src/cffi-1.3.0, inifile:
collected 1805 items / 1 skipped

D:/builds/mingw-packages/mingw-w64-python-cffi/src/cffi-1.3.0/c/test_c.py .........s..........................................................F.......sss.........Segmentation fault
~~~~

It's cool that the tests are mostly passing, but I would need to look into the failure and the segmentation fault before I felt this package is ready to be used in MSYS2.

You can see my progress so far on this branch:

https://github.com/DavidEGrayson/MINGW-packages/commits/py_cffi

I had to write a small patch to basically tell cffi to use the system's libffi (it wrongly assumes that any win32 system does not have a working libffi so it compiles one itself).  Then I had to write a patch to fix another small issue.

For what it's worth, my cffi PKGBUILD can successfully build all four packages and install them, and then if I run `pip install cryptography` after that, it tried to install the cryptograph pip and fails because of some MSVC-style compiler options that they are trying to use:

~~~~
$ pip install cryptography
....
    D:\msys64\mingw64\bin/x86_64-w64-mingw32-gcc.exe -shared -Wl,--enable-auto-image-base -pipe -s -s build/temp.mingw-2.7/build/temp.mingw-2.7/_openssl.o build/temp.mingw-2.7/build/temp.mingw-2.7/_openssl.def -LD:/msys64/mingw64/lib/python2.7/config -LD:/msys64/mingw64/lib -llibeay32 -lssleay32 -ladvapi32 -lcrypt32 -lgdi32 -luser32 -lws2_32 -lpython2.7 -o build/lib.mingw-2.7/cryptography/hazmat/bindings/_openssl.pyd /NXCOMPAT /DYNAMICBASE
    x86_64-w64-mingw32-gcc: error: /NXCOMPAT: No such file or directory
    x86_64-w64-mingw32-gcc: error: /DYNAMICBASE: No such file or directory
    error: command 'D:\\msys64\\mingw64\\bin/x86_64-w64-mingw32-gcc.exe' failed with exit status 1
~~~~

So we would need a MSYS2 package for the cryptography pip as well.

This is interesting work, but it's kind of discouraging.  I have heard that the CPython maintainers are not interested in accepting any patches to get MinGW working, and that is why the MSYS2 developers had to add [tons of patches to python2](https://github.com/Alexpux/MINGW-packages/tree/master/mingw-w64-python2) to get it to work.  I think that even with all these patches, it still doesn't work perfectly.  For example, when I use pip to uninstall a package (e.g. `pip uninstall pycparser`), it stalls forever and if I press Ctrl+C my MinTTY/Bash gets messed up.

Hey, why are we talking about the `cryptography` pip anyway?  I just downloaded the yotta pip and I see that these are its dependencies:

~~~~
semantic_version>=2.3.1,<3
requests>=2.5,<3
PyGithub>=1.25,<2
colorama>=0.3,<0.4
hgapi>=1.7,<2
Jinja2>=2.7.0,<3
PyCrypto>=2.5,<3
PyJWT>=0.3,<0.4
pathlib>=1.0.1,<1.1
~~~~

PyCrypto seems to install cleanly with `pip install PyCrytpo`.

**Comment by autopulated** [comment]
pycrypto also needs compilation, so I don't think it would help. Old versions of yotta used PyCrypto (>8m old at this point), we switched to cryptography because of ambiguous licensing, and installation challenges with pycrypto on different platforms.

**Comment by DavidEGrayson** [comment]
Sorry, I am not going to take charge of this anymore.  Yotta depends on a lot of things from the Python world and there are probably a lot of patches needed to those things because they assume that Python in Windows will use MSVC as the compiler instead of mingw-w64.  Any package that needs a patch should probably converted to be an MSYS2 package and installed that way instead of using pip.  To make the whole installation easy to understand, probably every dependency of yotta would need its own MSYS2 package; or else the user would have to be aware of which packages are installed with pip and which packages are installed with pacman and run multiple commands to install everything.  Also, yotta as a product is uncomfortably close to what I do in my real job.  However, if you want to take the initiative and get yotta working on MSYS2, I'd be happy to help with questions either here, or you could email me, or post on the #msys2 IRC channel on irc.oftc.net:6667.  In particular, I could help explain the process of [building a package](https://sourceforge.net/p/msys2/wiki/Contributing%20to%20MSYS2/) if you're not sure how that works.

**Comment by BlackstoneEngineering** [comment]
Fair enough, thank you for taking time to contribute and for the guidance. I will continue to keep an eye on MSYS2 as an option.

Go forth and build awesome!!

**Comment by BlackstoneEngineering** [comment]
Marking this as help wanted and wont fix for the time being, leaving it in place for future improvement.
